train/custom_datasets.py

```diff
@@ -29,172 +37,86 @@ def split_data(raw_data, input_file, percentage):
-
-    return text, input_ids, targets
-
-
-def prepare_messages_for_model(messages, tokenizer):
-    """Prepares a list of messages for the model by calling `prepare_message_for_model` function on each of them and
-    concatenating the returned input_ids and targets. Also, the function merges the text of the messages.
+def prepare_training_inputs(
+    messages: List[Dict],
+    tokenizer: Any,
+    padding: str = "max_length",
+    max_length: Optional[int] = None,
+    return_tensor: bool = True,
+    verbose=False,
+) -> Tuple[str, Dict]:
+    """This function is used for when you want to get a dictionary input for the model.forward.
+    The dictionary will contain: input_ids, attention_maks, labels.
+    labels is like input_ids except that content from user, system, function will be set as -100, only content from assistant remains
+
+    Args:
+        messages (List[Dict]): List of messages in openAI format (containing: role, content and function_call (optional))
+        tokenizer (Any): tokenizer from transformers
+        padding (str, optional): type of padding (longest, max_length), this is passed to tokenizer(). Defaults to "max_length".
+        max_length (Optional[int], optional): maximum number of tokens allowed in prompt. Defaults to None.
+        return_tensor (bool, optional): if true, the input_dic will be dictionary[str, Tensor] else dictionary[str, List[int]]. Defaults to True.
+        verbose (bool, optional): to print some useful information or not. Defaults to False.
+
+    Returns:
+        Tuple[str, Dict]: (final_prompt_str, inputs)
+            final_prompt_str: the final prompt to be used,
+            inputs: a dictionary containing: input_ids, attention_mask, labels. This will be used in model.forward(**inputs)
     """
-    all_texts = []
-    all_input_ids = []
-    all_targets = []
-
-    for message in messages:
-        text, input_ids, targets = prepare_message_for_model(message, tokenizer)
-        all_texts.append(text)
-        all_input_ids.append(input_ids.squeeze(0))
-        all_targets.append(targets.squeeze(0))
-
-    input_ids_tensor = torch.cat(all_input_ids, dim=-1)
-    targets_tensor = torch.cat(all_targets, dim=-1)
-    merged_text = "".join(all_texts)
-
-    prepared_input = tokenizer.prepare_for_model(
-        input_ids_tensor.tolist(),
-        return_tensors="pt",
-        padding="max_length",
-        truncation=True,
-        max_length=tokenizer.model_max_length,
+    # a dictionary mapping from token_id --> end_token
+    id_to_endtoken = get_token_id_to_end_token(tokenizer)
+    prompt_str = (
+        "system:\n"
+        + generate_schema_from_functions(functions=messages["functions"])
+        + "\nsystem:\nA chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions. The assistant calls functions with appropriate input when necessary\n"
```

Comment:
We should let the get_prompt_from_messages function to prepare the input. So, prepend this as a message dict to the original messages.

Reply:
with proper role and content etc